### ISSUE TITLE
[Feature request] Functions to read the contents of text and binary files

### ISSUE THREAD

**Issue by Rangi42** [gbdev/rgbds#933]
`INCLUDE` and `INCBIN` can include a file's contents directly as assembly code or data, but there are times when it would be nice to access a file's contents arbitrarily. For example, you might want to do something different depending on the size of the file; or reuse a tilemap with some offset added to all the bytes; or use part of a file depending on an rgbasm calculation; or so on. This currently has to be handled by including files generated with custom tools according to particular Makefile rules.

If we had arbitrary-length strings (#650) and arrays (#67), this could be accomplished with just two functions:

- `READFILE("file.txt")` would return a string of the file's characters
- `READBIN("file.dat")` would return an array of the file's bytes

For both functions, if the file doesn't exist they would abort with a "No such file" error, or under `-M` dependency-generating mode they would act consistent with `INCLUDE` and `INCBIN` after #903 gets resolved.

Note that `READFILE` would read the file's bytes as-is, which may not be valid UTF-8. It's already possibly to create a non-UTF-8 .asm file with arbitrary bytes in a string literal, so this is not a new issue. Some functions like `STRCAT` and `STRIN` wouldn't care about invalid UTF-8, others like `STRLEN` and `STRSUB` would print a "Invalid UTF-8 byte" error.

**Comment by Rangi42** [comment]
https://github.com/gbdev/rgbds/issues/67#issuecomment-966679756 discusses some possible syntaxes and issues with native arrays. If we ended up not adding them, another approach to reading binary file content would be to just read it as a string too, and just trust the user not to do anything with the string that expects valid UTF-8.

This would require adding a few byte-based functions (`STRBYTELEN` and `STRBYTESUB` at least) to use on "binary strings". String functions like `STRCAT` and `STRRPL` currently don't care about how a string is encoded, so they could be used on binary strings too...

...except that even they are implemented with C's own strings functions, so would behave incorrectly if the binary string has any $00 bytes. Even the ongoing PR #650 to have dynamic strings would probably still have that limitation.

(I'm also not really a fan of adding this kind of implicit limitation, where the notion of "valid string" vs "binary string" is all up to the user's understanding and has no type-level enforcement in the rgbasm language. It's also less flexible than arrays, which could store any numbers, not just bytes. But I'm bringing it up because arrays have issues of their own, and would be a large feature to add just for the same of `READBIN`.)

**Comment by Rangi42** [comment]
C++ `std::string` does support storing NUL bytes, so presumably if we used C++ std lib algorithms instead of C's string.h functions, they would work as expected for "binary strings".

```c++
#include <iostream>
#include <string>
int main() {
    std::string s("hello");
    std::cout << s.size() << " " << s << std::endl;
    s[2] = '\0';
    std::cout << s.size() << " " << s << std::endl;
    return 0;
}
```
```
5 hello
5 helo
```

**Comment by Rangi42** [comment]
Currently pokegold uses a custom tool to measure the size of a 5x5, 6x6, or 7x7 PNG image and encode that as $55, $66, or $77. This could replace its `INCBIN "front.dimensions"` lines with a simpler `INCDIMENSIONS "front.2bpp"` macro (named that way so `scan_includes` could easily be adapted to handle it):

```
MACRO INCDIMENSIONS
    def n = STRBYTELEN(READBIN(\1)) ; or STRLEN(READFILE(\1)) pending #938
    if n == 5**2 tiles
        db $55
    elif n == 6**2 tiles
        db $66
    elif n == 7**2 tiles
        db $77
    else
        fail "Invalid front pic size for \1"
    endc
ENDM
```

**Comment by Rangi42** [comment]
`READFILE` could also take optional *start* and *length* arguments like `INCBIN`. This would be more efficient for use cases that only need part of a file, and would let pokecrystal do this:

```
MACRO INCDIMENSIONS
    pushc
    setcharmap ascii
    DEF png_ihdr_width EQUS READFILE(\1, 16, 4)
    DEF n = STRSUB("{png_ihdr_width}", 1, 1) << 24 | \
            STRSUB("{png_ihdr_width}", 2, 1) << 16 | \
            STRSUB("{png_ihdr_width}", 4, 1) <<  8 | \
            STRSUB("{png_ihdr_width}", 5, 1)
    if n == 5 * TILE_LENGTH
        db $55
    elif n == 6 * TILE_LENGTH
        db $66
    elif n == 7 * TILE_LENGTH
        db $77
    else
        fail "Invalid front pic size for \1"
    endc
    PURGE png_ihdr_width
    popc
ENDM
```

**Comment by Rangi42** [comment]
We could compare this proposal with WLA-DX's `.FOPEN`, etc directives.

```asm
.fopen "data.bin" fp
.fsize fp t
.repeat t
.fread fp d
.db d+26
.endr
.undefine t, d
```

**Comment by ISSOtm** [comment]
I don't like exposing arbitrary file I/O because it's just complicated:

- Try eagerly reading all of the file, and potentially OOM for large files that could've been processed piecemeal, or block infinitely from e.g. pipes;
- Try exposing piecemeal read functions (à la WLA-DX), and shift all of the complexity onto the user, and still occasionally block infinitely.

...so for the time being, I still prefer people building ad-hoc tools and solutions when such needs arise. Please understand that I'm not trying to be a grouch raining on everyone's parade, but also reining in what I fear to be scope creep.

I am however leaving this issue open so that people can still voice any support, use cases, or design suggestions.

**Comment by Rangi42** [comment]
> Try eagerly reading all of the file, and potentially OOM for large files that could've been processed piecemeal, or block infinitely from e.g. pipes;

Don't we have exactly this theoretical problem already, with `INCBIN`? I'd expect a `READFILE` function to work just like that, except putting the data in an array, not directly in the ROM.

**Comment by ISSOtm** [comment]
In theory yes, but not really because `INCBIN` means that all of the file's contents must be dumped into the ROM anyway, and we *have* to allocate that space anyway for a few reasons.

**Comment by Rangi42** [comment]
Reading *binary* files is full of complications. We can't just read them into `EQUS` strings. Many operations on strings expect valid UTF-8 (including `db`, since it goes by charmap entries, not individual bytes), so we would need either a new set of bytewise operations on strings (including "`dbb`", `BYTELEN`, `BYTESLICE`, etc) or a whole new "array" type separate from `EQUS` (which gets way more complicated as we start to want user-defined arrays, utility functions to index/append/reverse/sort, string↔array functions, etc).

On the other hand, we should be able to read *text* files into strings. Invalid UTF-8 could either be a total error and evaluate to an empty string, or substitute or omit the invalid character and continue processing. (The latter would be more like current "invalid UTF-8" handling elsewhere.)

ISSOtm brings up a problem with this, where `READFILE("/dev/zero")` or some unlimited-length file could hang or OOM. To avoid this, we could *require* a max-length argument like `READFILE("/dev/zero", $ffff)`, similar to how `INCBIN` can take an offset and length. (`READFILE` does not need an offset, because most of the time it would be zero so `READFILE("/dev/zero", 0, $ffff)` would be boilerplate, and if you really need to slice off the start of a string you already have `STRSLICE`.)

**Comment by Rangi42** [comment]
So, I still like the idea of reading whole text files into `EQUS` constants, like `DEF dialog EQUS READFILE("dialog.txt")` (or at least `DEF dialog EQUS READFILE("dialog.txt", $ffff)` to cap the max length read).

But binary files don't have to follow the same paradigm! They could be their own whole new kind of symbol, a *blob* (like in SQL). The only expected operations on blobs would be syntax to get the *length* and the *Nth byte*. There'd be no expectation of storing full-size 32-bit numbers, just bytes; no expectation of appending or reversing or splitting or shuffling them like an array; no expectation of defining them yourself except from reading a binary file.

(Yes, it should be fine to have a new kind of symbol that isn't a number or string -- macro names already are that.)

Syntax could be bikeshedded:

- `DEFBIN tilemap, "tilemap.bin", $ffff` ✅ (matches `INCBIN`)
- `DEFBLOB tilemap = "tilemap.bin", $ffff`
- `DEF tilemap EQUB "tilemap.bin", $ffff`
- `DEF tilemap BINBLOB "tilemap.bin", $ffff`
- plenty of options

Likewise syntax for the length and Nth byte:

- `BINLEN(tilemap)` and `BINBYTE(tilemap, idx)` ✅ (matches `STRLEN` and `STRBYTE`)
- `BINLEN(tilemap)` and `BINVAL(tilemap, idx)`
- `BLOBLEN(tilemap)` and `BLOBVAL(tilemap, idx)`
- `tilemap[#]` and `tilemap[idx]`
- `(tilemap)` (or any other manner of treating `tilemap` as a number, e.g. `+tilemap` or `tilemap - 0`) and `tilemap[idx]`
- plenty of options

If someone really wants to manipulate a binary file as a dynamic array, it wouldn't be hard:

```asm
FOR idx, BINLEN(tilemap)
    DEF tilearr#{d:idx} EQU BINVAL(tilemap, idx)
ENDR
DEF tilearr#LEN EQU idx
```

Then `tilearr` would be compatible with the arrays.inc macros commented in #67.

Problems this solves:

- Allows text files to go in `EQUS` constants, making use of all the functions and features we already have for working with strings.
- Allows binary files to be read at all, even though they're not valid encoded string text.
- Avoids the temptation to add a bunch of built-in array functions and syntax.

**Comment by Rangi42** [comment]
A Boolean `FEXISTS` function (like WLA-DX) would probably be helpful here. It would also complement `rgbasm -MC`, so if you ever do want to use `-MC` for 99% of your program but 1% of `INCLUDE`s would affect subsequent compilation, you can handle the case where they haven't been generated yet. (But see the discussion in #724 first! And avoid TOCTTOU errors if possible.)
